### ISSUE TITLE
Modify RDL Node Class to Remove Socket/Core Requirements

### ISSUE THREAD

**Issue by lipari** [flux-framework/flux-sched#200]
The 7bc281217cd9dc0d14274850de561be74d2b2f67 commit introduced the flow graph support to the resrc code base.  When a non-default hierarchy that includes a Node object is requested in an RDL conf file, the existing Node class construction requires the presence of at least one socket and core.  This is not needed or desired when the node is the leaf of a power distribution hierarchy.  This issue calls for the modification of the RDL's Node class to not require Socket or Core children.  This should be made optional, as there might be a means to measure and report power/energy consumption at the socket or core level.

**Comment by dongahn** [comment]
Do we already have the base class of a resource in RDL? If so, a node resource can be a derived class of the base class and contains at least 1 core and 1 socket. And a flow resource is another derived class of the base or just an instantiation of the base. 

BTW, I forgot, are the core and socket modeled as a resources of properties?

**Comment by grondo** [comment]
Yes, Node, Socket, Core are all derived from base Resource class.

I was going to suggest something similar, like a PSU (or flow as you suggest) class, but I'm not sure how to best relate the PSU to a Node, there might not be support for that directly in current RDL config language. How it should work might depend on the underlying resrc implementation.

**Comment by dongahn** [comment]
@grondo: completely agreed that would be the ideal implementation.

@lipari can chime in, but `resrc` introduces the power flow resource for each physical node . And this power flow resource has a pointer to the actual physical node to represent sort of a "resource_of" relation. (i.e., electric current flow is a resource of this physical node.) 

And this power flow node and the physics node association is done through the resource indexing support. This also means, we need to parse the physical hierarchy first before moving on to the next hierarchy.. which is less than ideal but should be manageable. 


I think, as far as RDL allows us to generate this power flow resource for each physical compute node in the power hierarchy, `resrc` should be able to do the rest of the mapping w/ the physical  node.

**Comment by lipari** [comment]
I think you're making this much more complicated than it needs to be.  I do not need a redesign of the RDL to accommodate a flow graph.  The hierarchy loading works just fine.  I just need to stop at the Node resource and not automatically require/populate sockets and cores.

**Comment by grondo** [comment]
@lipari, I guess I'm lost, I would think when populating a power hierarchy you would want a different resource type than a `Node` which is a compute node with Sockets, Cores and Memory.

Creating a new type subclass of Resource is trivial, not much more difficult than removing the requirement that compute `Node` has at least one Socket with at least one Core.

Possibly I need to look at the current use case.

**Comment by grondo** [comment]
Ok, looking at the example I understand a bit better. What you actually want is to be a create a list of references to Nodes that are already populated in the physical hierarchy. The current usage creates duplicate resources in the RDL for each separate instatiation of a `Node{ }` type. I think this is a case where the design the RDL fell short.

Can you remind me what the "size" parameter is for?

If it is sufficient to allow these duplicate Nodes to be created without socket/core children, then that will be an easy workaround for now. However, in my opinion it would be much cleaner to not overload the `Node` type like this, and instead create a `NodePSU` or `PowerFlow` type with similar constructor to the `Node` type, but doesn't support Sockets/Cores/Memory. This could be thought of as pointing to the Node by reference  or similar. It should not be any extra work from your end, except types would need to be updated when walking the power hierarchy.

Something similar could be done for the bandwidth graph.

**Comment by lipari** [comment]
As just discussed, @grondo's plan for the power and bandwidth node types is the way to go.

**Comment by grondo** [comment]
Another idle thought comes to mind here. We *could* also create a `PDU` type that takes a list of hostlist, "size" pairs as one of its arguments (might come up with a better name than size if it is PDU specific argument). This could shorten declaration of the power flow graph to something like
```lua
Hierarchy "power" {
￼     Resource{ "powerpanel", name="ppnl", size = 1000, children = {
￼       PDU{ id = "1", size = 500, nodes = { { nodes = "hype[201-299]", size = 10 } } },
       PDU{ id = "2", size = 500, nodes = { { nodes = "hype[300-354]", size = 10 } } },
       }
     }
 }
   
```

The PDU initializer would add its own children based on the nodes hostlist, as the Node initializer does for Sockets.

**Comment by lipari** [comment]
First off, "amps" would be the appropriate term to express current flow.  We should probably use "amps" even though it's a "power" hierarchy, because power is simply current * voltage.  PDU could serve as the parent of node, but there could be cases where there's an intervening distribution device.  I can imagine a 8 nodes in a chassis enclosure with its own power supply and 4 chassis connected to a PDU.

**Comment by garlick** [comment]
I would be careful about using amps as a standin for power, since voltage is not constant in a real system, and power is not normally calculated as amps * volts in an AC circuit.

**Comment by grondo** [comment]
> I can imagine a 8 nodes in a chassis enclosure with its own power supply and 4 chassis connected to a PDU

In that case in typical OOP fashion we could have a base `PDU` class and derive a `simplePDU` (or insert better name here) that takes the nodelist argument. You could also get very specific and add a `hypePDU` directly to the config file, in which case the `nodes` list might be simplified. I'm just throwing these instructive ideas out there, as that was the purpose of choosing a class system for resource objects (whatever makes your life easier at this point, though)

**Comment by grondo** [comment]
Good points about the units for power. I'm not sure what the meaning of `size` in the current (sorry no pun intended) implementation is, so I can't suggest an alternative, but I think `size` will probably be confusing either way. Perhaps even something like `capacity` or `limit` would be be better?

Eventually admins may want to configure these resources "as they stand" in real life, so a PDU may have a rating or limit, which the configurer would enter as is,  and the `PDU` class initializer would turn that into watts or whatever the scheduler might need. (That is how I envision a type based configuration language working anyway)

**Comment by dongahn** [comment]
Yeah, I agree we should try to introduce some techniques that allows us to specify various hierarchy in a condensed form. @SteVwonder's IO hierarchy had very large entries, which looked prone to errors... One technique that might be helpful is to allow the id of a nested resource to be a function of the id of its nesting resource. Generally speaking, I like the @grondo's PDU specification...

**Comment by dongahn** [comment]
> Good points about the units for power. I'm not sure what the meaning of size in the current (sorry no pun intended) implementation is, so I can't suggest an alternative, but I think size will probably be confusing either way. Perhaps even something like capacity or limit would be be better?

> Eventually admins may want to configure these resources "as they stand" in real life, so a PDU may have a rating or limit, which the configurer would enter as is, and the PDU class initializer would turn that into watts or whatever the scheduler might need. (That is how I envision a type based configuration language working anyway)

We may also want to look at what quantify that CPU's Power capping technologies using. For physical PDUs and etc, it seems the quantities should match what operators are using... But for node power capping, the resource manager will have to set the limit, so it seems we may want to match it w/ current technologies... I will take a look.

**Comment by lipari** [comment]
Yes, we have two units that we could use to model electrical flow.  If we are looking to model current flow across physical wiring at constant voltage, then `amps` is the right term.  But if we take a bigger picture and model power from the power coming into the center which could be at a higher voltage, then `watts` is better.  Also, modeling a "power envelope", `watts` is more appropriate.  @tpatki, what do you think?

**Comment by dongahn** [comment]
@tpatki can chime in. 

My brief google search shows that changes are they use Watts. E.g., https://www.kernel.org/doc/Documentation/power/powercap/powercap.txt says they have parameters like `constraint_0_power_limit_uw` This probably reads "for some power domain of a microprocessor, the power is capped to k microwatts.

**Comment by lipari** [comment]
Ok, I'm convinced.  Unless @tpatki has better insight, let's go with `watts`.  And given that our resrc model has integer `size` values, we should probably make `uwatts` the choice.

**Comment by grondo** [comment]
@lipari, I was suggesting a new attribute other than `size` and there is no reason that value can't be a float or double.

**Comment by lipari** [comment]
Alright.  The resrc code uses size_t size everywhere, so this would be a big change if propagated throughout all the resrc code.  I'll probably just scale up the float uwatt to a size_t value for the time being.

**Comment by tpatki** [comment]
I think Watts is appropriate for power. We may also need to think of energy consumption. There is a notion of instantaneous and average power/energy as well.

**Comment by tpatki** [comment]
That being said CPU frequency is an important concern too. 

I believe for now we are assuming there will be a way to monitor power instantaneously? And that we can use a feedback loop to set power cap/CPU frequency given the scheduling policy. This will require some interfacing with low-level power management tools (such as libmsr/msr-safe on intel machines), which differ based on architecture.

**Comment by lipari** [comment]
The simplest model we're looking to support with the power hierarchy is flow.  We want to schedule flow such that jobs can only be scheduled when there is adequate power available.  In this simple model, the job specifies a power requirement (or gets a default), and we stop scheduling more jobs when all the available power has been allocated.

Once that is done, we can consider another layer of complexity whereby dynamic power envelopes are applied, actual power consumed is measured, and cpu frequencies are adjusted to meet power consumption demands.  I'd like to start of with the simple case first, making sure we don't lock in a design that prevents enhancements that support power capping and frequency adjustments in the future.

**Comment by dongahn** [comment]
> I think Watts is appropriate for power. We may also need to think of energy consumption. There is a notion of instantaneous and average power/energy as well.

Where are these different notions used? 

I vaguely thought that a scheduler would schedule and allocate power and then the amount of the execution time a job runs on that allocation would determine energy consumption... Whether you want to optimized for energy saving or performance under a power cap would demand different scheduling policies...

A bit more details on where instantaneous vs average power/energy are used should be very useful at this point. 

> And that we can use a feedback loop to set power cap/CPU frequency given the scheduling policy. 

In order for this to work, I think we probably need more advanced techniques than what we currently have today: 1) we use the Flux instance to set the overall power cap and relies on a underlying a power-aware runtime system to take care of dynamic cap/cpu frequency settings within this limit. I think doing this at the individual node level should be straightforward. 2) if you want to do this at higher than node-level, you may need this logic implemented in a form of dynamic scheduler so that power can be redistributed?

**Comment by grondo** [comment]
> Alright. The resrc code uses size_t size everywhere, so this would be a big change if propagated throughout all the resrc code. I'll probably just scale up the float uwatt to a size_t value for the time being.

Ah, ok, I understand now. The resrc objects must use `size` to encode any amount/limit/level/etc.? 

We could convert internally to strict integer size within the RDL configuration itself, or you could scale `watts` properties for these objects directly in the resrc RDL reader code, up to you.

**Comment by lipari** [comment]
I guess, let's put the scaling to integer size in the RDL configuration for now.  This keeps the resrc reader code uniformly dealing with ints.

**Comment by morrone** [comment]
Are you saying that the amps/watts/whatever value is literally stored in
a variable declared as a size_t?  size_t  is meant to be used for
representing the size of objects in memory in C/C++, I would not think
that it should be used for storing values that represent real-world
physical values.

Chris

**Comment by lipari** [comment]
Yes, the resrc code is consistent with the RDL library (rdl.h) that it uses to read the resource config.

**Comment by grondo** [comment]
> Yes, the resrc code is consistent with the RDL library (rdl.h) that it uses to read the resource config.

Yeah, that is my fault, I don't know what made me choose `size_t`. It was originally meant to represent a count for a pool of individually indistinct resources for accounting available/inuse/etc. It probably should have been `uint64_t` or even just 32 bit unsigned.

**Comment by grondo** [comment]
closed by #440
